src/main/java/com/commercetools/sync/products/helpers/VariantReferenceResolver.java

```diff
@@ -0,0 +1,191 @@
+        final Spliterator<JsonNode> attributeReferencesIterator = attributeDraftValue.spliterator();
+        final List<CompletableFuture<JsonNode>> referenceResolutionFutures =
+            StreamSupport.stream(attributeReferencesIterator, false)
+                         .filter(Objects::nonNull)
+                         .filter(reference -> !reference.isNull())
+                         .map(reference -> {
+                             if (isProductReference(reference)) {
+                                 return getResolvedIdFromKeyInReference(reference)
+                                     .thenApply(productIdOptional ->
+                                         productIdOptional.map(this::createProductReferenceJson)
+                                                          .orElse(reference));
+                             }
+                             return CompletableFuture.completedFuture(reference);
+                         })
+                         .map(CompletionStage::toCompletableFuture)
+                         .collect(Collectors.toList());
+        return CompletableFuture
+            .allOf(referenceResolutionFutures.toArray(new CompletableFuture[referenceResolutionFutures.size()]))
+            .thenApply(voidResult -> {
+                final List<JsonNode> resolvedProductReferences =
+                    referenceResolutionFutures.stream()
+                                              .map(CompletionStage::toCompletableFuture)
+                                              .map(CompletableFuture::join)
+                                              .collect(Collectors.toList());
+                return AttributeDraft.of(attributeDraft.getName(), resolvedProductReferences);
+            });
+    }
+
+    static boolean isProductReference(@Nonnull final JsonNode referenceValue) {
+        return getReferenceTypeIdIfReference(referenceValue)
+            .map(referenceTypeId -> Objects.equals(referenceTypeId, "product"))
+            .orElse(false);
+    }
+
+    private static Optional<String> getReferenceTypeIdIfReference(@Nonnull final JsonNode referenceValue) {
+        final JsonNode typeId = referenceValue.get("typeId");
+        return typeId != null ? Optional.ofNullable(typeId.asText()) : Optional.empty();
+    }
+
+    CompletionStage<Optional<String>> getResolvedIdFromKeyInReference(@Nonnull final JsonNode referenceValue) {
+        final JsonNode idField = referenceValue.get("id");
+        return idField != null
+            ? productService.fetchCachedProductId(idField.asText())
+            : CompletableFuture.completedFuture(Optional.empty());
+    }
+
+    private JsonNode createProductReferenceJson(@Nonnull final String productId) {
+        final ObjectNode productReferenceJsonNode = JsonNodeFactory.instance.objectNode();
+        productReferenceJsonNode.put("id", productId);
+        productReferenceJsonNode.put("typeId", "product");
+        return productReferenceJsonNode;
+    }
+}
+
```

Comment:
i found 3 duplications in this file, created a patch and sent you. Pls, review and tell me WDYT (merged most of `allOf/map/stream` futures)

Reply:
thanks for the proposal f996ed7